runtime/parachains/src/scheduler.rs

```diff
@@ -246,42 +236,63 @@ impl<T: Config> Pallet<T> {
 		(concluded_paras, timedout_paras)
 	}
 
-	/// Note that the given cores have become occupied. Behavior undefined if any of the given cores were not scheduled
-	/// or the slice is not sorted ascending by core index.
-	///
-	/// Complexity: O(n) in the number of scheduled cores, which is capped at the number of total cores.
-	/// This is efficient in the case that most scheduled cores are occupied.
+	/// Note that the given cores have become occupied. Update the claimqueue accordingly.
 	pub(crate) fn occupied(
 		now_occupied: BTreeMap<CoreIndex, ParaId>,
 	) -> BTreeMap<CoreIndex, PositionInClaimqueue> {
 		let mut availability_cores = AvailabilityCores::<T>::get();
 
-		let pos_mapping = now_occupied
-			.into_iter()
-			.flat_map(|(core_idx, para_id)| match Self::remove_from_claimqueue(core_idx, para_id) {
-				Err(_) => None, // TODO: report back?
-				Ok((pos_in_claimqueue, assignment)) => {
-					// is this correct?
-					availability_cores[core_idx.0 as usize] = assignment.to_core_occupied();
-
-					Some((core_idx, pos_in_claimqueue))
+		let mut pos_mapping = BTreeMap::new();
+		for (core_idx, para_id) in now_occupied.iter() {
+			match Self::remove_from_claimqueue(*core_idx, *para_id) {
+				Err(e) => {
+					log::debug!(
+						target: LOG_TARGET,
+						"[occupied] error on remove_from_claimqueue {}",
+						e
+					);
 				},
-			})
-			.collect();
+				Ok((pos_in_claimqueue, pe)) => {
+					availability_cores[core_idx.0 as usize] = CoreOccupied::Paras(pe);
+					pos_mapping.insert(*core_idx, pos_in_claimqueue);
+				},
+			}
+		}
+
+		// For now, we drop the first entry of the claimqueue for each CoreIndex that did not get occupied
+		let occupied_cores = now_occupied.into_keys().collect();
+		let all_cores: BTreeSet<CoreIndex> =
+			(0..availability_cores.len() as u32).map(CoreIndex::from).collect();
+		let cores_not_occupied = all_cores.difference(&occupied_cores);
+		for core_idx in cores_not_occupied {
+			Self::drop_claimqueue_front_for_unoccupied_core(core_idx);
+		}
 
 		AvailabilityCores::<T>::set(availability_cores);
 
 		pos_mapping
 	}
 
+	// TODO: tests
+	fn drop_claimqueue_front_for_unoccupied_core(core_idx: &CoreIndex) {
```

Comment:
This name reflects the calling side, not what the function actually does. Better name maybe something like `drop_claimqueue_front_refill_back_from_provider` or shorter `drop_claimqueue_front_refill` ... with doc explaining what it does.

Reply:
This is one of the functions that will be removed when TTL is in as this was only our first attempt to solve the issue of faulty paras in the claimqueue not getting dropped.